features/networking/ovn.feature

```diff
@@ -0,0 +1,224 @@
+    """
+    And admin waits for all pods in the project to become ready up to 60 seconds
+
+
+  # @author rbrattai@redhat.com
+  # @case_id OCP-26091
+  @admin
+  @destructive
+  Scenario: New corresponding raft leader should be elected if SB db or NB db on existing master is crashed
+    Given admin uses the "openshift-ovn-kubernetes" project
+    When I store the ovnkube-master "south" leader pod in the clipboard
+    Then the step should succeed
+    When the OVN "south" database is killed on the "<%= cb.south_leader.node_name %>" node
+    Then the step should succeed
+
+    And I wait up to 30 seconds for the steps to pass:
+    """
+    When I store the ovnkube-master "south" leader pod in the :new_south_leader clipboard
+    Then the step should succeed
+    And the expression should be true> cb.south_leader != cb.new_south_leader
+    """
+    And admin waits for all pods in the project to become ready up to 60 seconds
+
+    When I store the ovnkube-master "north" leader pod in the clipboard
+    Then the step should succeed
+    When the OVN "north" database is killed on the "<%= cb.north_leader.node_name %>" node
+    Then the step should succeed
+
+    And I wait up to 30 seconds for the steps to pass:
+    """
+    When I store the ovnkube-master "north" leader pod in the :new_north_leader clipboard
+    Then the step should succeed
+    And the expression should be true> cb.north_leader != cb.new_north_leader
+    """
+    And admin waits for all pods in the project to become ready up to 60 seconds
+
+  # @author rbrattai@redhat.com
+  # @case_id OCP-26138
+  @admin
+  @destructive
+  Scenario: Inducing Split Brain in the OVN HA cluster
+    Given admin uses the "openshift-ovn-kubernetes" project
+    When I store the ovnkube-master "south" leader pod in the clipboard
+    Then the step should succeed
+
+    Given I store the masters in the clipboard excluding "<%= cb.south_leader.node_name %>"
+    And I use the "<%= cb.nodes[0].name %>" node
+    # don't block all traffic that breaks etcd, just block the OVN ssl ports
+    When I run commands on the host:
+      | iptables -t filter -A INPUT -s <%= cb.south_leader.ip %> -p tcp --dport 9643:9644 -j DROP |
+    Then the step should succeed
+
+    And I wait up to 30 seconds for the steps to pass:
+    """
+    When I store the ovnkube-master "south" leader pod in the :new_south_leader clipboard
+    Then the step should succeed
+    And the expression should be true> cb.south_leader == cb.new_south_leader
+    """
+    When I run commands on the host:
+      | iptables -t filter -D INPUT -s <%= cb.south_leader.ip %> -p tcp --dport 9643:9644 -j DROP |
+    And I wait up to 30 seconds for the steps to pass:
+    """
+    When I store the ovnkube-master "south" leader pod in the :new_south_leader clipboard
+    Then the step should succeed
+    And the expression should be true> cb.south_leader == cb.new_south_leader
```

Comment:
still getting `RuntimeError: expression "cb.south_leader == cb.new_south_leader" returned non-positive status: false
`

so need to re-think this test with respect to Raft behavior.  I'll remove this test from this PR and submit it later.

Reply:
@anuragthehatter figured out how to track leader votes using `ovsdb-tool`

```
oc  -n openshift-ovn-kubernetes exec ovnkube-master-vgtss -c sbdb -- ovsdb-tool show-log /etc/ovn/ovnsb_db.db | grep -B1 -e vote: -e leader:
```

Term is the election cycle.
```
 term: 2
 vote: fa33
--
 term: 2
 leader: fa33
```

Reply:
`for f in $(oc get pod -n openshift-ovn-kubernetes -l app=ovnkube-master -o jsonpath={.items[*].metadata.name})  ; do echo -e "\n${f}" ; for c in nb sb ; do oc  -n openshift-ovn-kubernetes exec -c "${c}db" ${f} -- ovsdb-tool show-log /etc/ovn/ovn${c}_db.db | grep -B1 -e vote: -e leader: -e prev_term: | tee votes-${f}-${c} ; done ; done`

Reply:
hmm, do you mean that the same leader can be re-elected post SB db crash on itself? Which case is references here and which particular line. Sorry lines quoted here doesn't align with what we see under "view file"

Reply:
Ah its split brain, never mind. Yea, we can track that separate. All i know right now is there will be a re-election if we block traffic on leader->follower. On re-election, either that follower on which the traffic was blocked will become leader and some other follower on a 3 node cluster. I believe ```expression "cb.south_leader == cb.new_south_leader"``` will always return false due to mismatch in leaders in this case as there will be new leader post traffic blocking. We can track this in detail later

Reply:
Yep `expression "cb.south_leader == cb.new_south_leader"` is false.  But I also realized that we aren't waiting for anything to time out after dropping the traffic, so maybe it was passing because we didn't wait for the leader election.

I was going to try to look up what timeout is for triggering a new election after losing contact with the leader.  I think you asked that before somewhere.

The actual test just calls for checking for a new leader, so we don't need to compare old leader with new leader, but we could compare once we figure it out.  Probably should wait >30 seconds then make sure the `new leader != old leader`.

Reply:
Make sense, thanks

Reply:
well it's more complicated.  Once we split, we technically have two leaders, the isolated node and the original leader.  Thus I had to modify the leader detector step to specify an individual node on which to check the leader.  It looks like the failures were that we were checking the leader on the isolated node, and the isolated node thinks it is it's own leader, so the leader comparison failed.  The original leader should stay the same because there aren't enough votes to change.

But then that implies more verification steps, after we remove the DROP rule we should verify that the isolated node re-joins and accepts the original leader.

I have an update, but I'm not sure on the OVN timing, should we wait for the isolated node to notice it is dropped, and how long should we wait after removing the DROP for OVN to re-converge.